numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,199 @@
+                             NPY_ITER_REFS_OK |
+                             NPY_ITER_ZEROSIZE_OK |
+                             NPY_ITER_GROWINNER,
+                        NPY_KEEPORDER, NPY_NO_CASTING,
+                        NULL);
+    // Making sure the iterator is deallocated when the function returns, with
+    // or w/o an exception
+    auto iter_dealloc = finally([&]() { NpyIter_Deallocate(iter); });
+    if (iter == NULL) {
+        return Py_None;
+    }
+
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        return Py_None;
+    }
+    dataptr = NpyIter_GetDataPtrArray(iter);
+    strideptr = NpyIter_GetInnerStrideArray(iter);
+    innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    // release the GIL
+    PyThreadState *_save;
+    _save = PyEval_SaveThread();
+    // Making sure the GIL is re-acquired when the function returns, with
+    // or w/o an exception
+    auto grab_gil = finally([&]() { PyEval_RestoreThread(_save); });
+    // first we put the data in a hash map
+    do {
+        char* data = *dataptr;
+        npy_intp stride = *strideptr;
+        npy_intp count = *innersizeptr;
+
+        while (count--) {
+            hashset.insert(*((T *) data));
+            data += stride;
+        }
+    } while(iternext(iter));
+
+    // then we iterate through the map's keys to get the unique values
+    T* res = new T[hashset.size()];
+    auto it = hashset.begin();
+    size_t i = 0;
+    for (; it != hashset.end(); it++, i++) {
+        res[i] = *it;
+    }
+
+    // does this need to have the same lifetime as the array?
+    npy_intp dims[1] = {(npy_intp)hashset.size()};
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    NPY_ALLOW_C_API;
+    Py_INCREF(descr);
+    PyObject *res_obj = PyArray_NewFromDescr(
+        &PyArray_Type,
+        descr,
+        1, // ndim
+        dims, // shape
+        NULL, // strides
+        res, // data
+        // This flag is needed to be able to call .sort on it.
+        NPY_ARRAY_WRITEABLE, // flags
+        NULL // obj
+    );
+    NPY_DISABLE_C_API;
+    return res_obj;
+}
+
+
+// this map contains the functions used for each item size.
+typedef std::function<PyObject *(PyArrayObject *)> function_type;
+std::unordered_map<int, function_type> unique_funcs = {
+    {NPY_BYTE, unique<npy_byte>},
+    {NPY_UBYTE, unique<npy_ubyte>},
+    {NPY_SHORT, unique<npy_short>},
+    {NPY_USHORT, unique<npy_ushort>},
+    {NPY_INT, unique<npy_int>},
+    {NPY_UINT, unique<npy_uint>},
+    {NPY_LONG, unique<npy_long>},
+    {NPY_ULONG, unique<npy_ulong>},
+    {NPY_LONGLONG, unique<npy_longlong>},
+    {NPY_ULONGLONG, unique<npy_ulonglong>},
+    {NPY_INT8, unique<npy_int8>},
+    {NPY_INT16, unique<npy_int16>},
+    {NPY_INT32, unique<npy_int32>},
+    {NPY_INT64, unique<npy_int64>},
+    {NPY_UINT8, unique<npy_uint8>},
+    {NPY_UINT16, unique<npy_uint16>},
+    {NPY_UINT32, unique<npy_uint32>},
+    {NPY_UINT64, unique<npy_uint64>},
```

Comment:
You could actually add datetime by using `int64/uint64` as the function, since the NaT values don't matter for "unique".

Reply:
A note for other reviewers: Floats are currently left out here to avoid having to deal with a custom hash/equal implementation (due to NaN values).